bpf:wireguard: delivery host packets to bpf_host for ingress policies

smagnani96 opened the pull request:
This effectively restores HostFw for WireGuard.
Prior to this, we were erroneously always returning to the stack all
packets destined for local host, skipping the HostFw policies if enabled.

With this patch, packets for local host will always be delivered to
cilium_net@egress, similarly to what we do in bpf_overlay after decap.
With HostFw enabled, the to-host program in `cilium_host@ingress` will
then enforce policies.

This patch does not affect packets for local endpoint:
* With BPF Host Routing: will be directly delivered to the pod, tail
  calling into its ep->lxc_id function to enforce ingress policies.
* Without BPF Host Routing: will return to stack, which then goes to its
  to-container installed program to match ingress policies.

Trying to pull-in the whole `host_firewall.h` and `policy.h` would require
to set bpf_wireguard similarly as we do for bpf_host, meaning assigning
an endpoint ID, otherwise host policies would block all host related
packets (ep id == 0). For this reason, we decide here to go through cilium_host.

From a bpf test perspective:
* no changes for packet to/from local endpoint INGRESS/EGRESS
* no changes for packet from local host EGRESS
* packet to local host INGRESS: differently than bpf_host, in WireGuard
                                we always redirect to cilium_host@ingress.

While fixing this bits, let's move the superfluous `revalidate_data`
post NodePort inside the NodePort code, as not needed otherwise.

smagnani96 commented on bpf/bpf_wireguard.c:
```diff
@@ -209,42 +225,31 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 identity, __s8 *ext_err __maybe_unused
-	return TC_ACT_OK;
-#endif
+	if (!is_defined(ENABLE_HOST_ROUTING))
+		return CTX_ACT_OK;
```
I tried on a local kind native routing with WG + Node Encryption + HostFW + CCNP/CNP:
* CNP:  for a local pod I always see CNP being applied.
* CCNP: I followed https://docs.cilium.io/en/latest/security/host-firewall/ to enable a simple host firewall rule and:
  * c/c:main -> we're NOT enforcing, we're always letting the packet go through the stack (TC_ACT_OK) after from-wireguard
  * this PR -> I can see ingress policy verdict drop for that host-to-remote-host ping.

HostRouting was always enabled, I did not test it with LegacyRouting: when we were attaching `bpf_host` in cilium_wg0, after hostFW enforcement we were always returning CTX_ACT_OK in legacy mode, so the behavior should be the same.

Given in this PR we're always forwarding host-destined packets to `cilium_net`, I don't understand how endpoint routes should affect this here: 🤔 
1. Are you referring to having something similar to what we do for [decrypted IPSec](https://github.com/cilium/cilium/blob/main/bpf/lib/ipsec.h#L170-L174) in `bpf_network`?
2. I don't see endpoint routes being used in bpf_overlay, or am I missing something?

smagnani96 commented:
fyi: given I'll backport this to v1.18 once finished, bpf tests have not been written in scapy-like language.

julianwiedmann commented:
For the last patch:

> Given we've done a legimitate lookup, let's make use of that identity.

I believe that's not accurate - `bpf_wireguard` currently still respects `secctx_from_ipcache`, and only resolves the source identity when BPF Host-Routing is enabled.

So if we now want to always forward an identity, then we also need to unconditionally resolve it (which feels like a reasonable choice to me).

julianwiedmann commented on bpf/bpf_wireguard.c:
```diff
@@ -95,33 +108,20 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 identity, __s8 *ext_err __maybe_unused
 
-#ifndef ENABLE_HOST_ROUTING
-	return CTX_ACT_OK;
-#endif
```
I believe from a netfilter-symmetry perspective, we need to step carefully here.

In `bpf_overlay` it's easy - we expect that both legs of the connection flow via the overlay network. And so packets by the local pod will get redirect to `cilium_vxlan` immediately, with no netfilter traversal. And therefore we can also unconditionally bypass netfilter for packets coming out of the overlay, towards the local pod.

But with Wireguard we do have the case (when BPF Host Routing is disabled) where a local pod's traffic first passes through netfilter, and only in `to-netdev` gets redirected to `cilium_wg`. And so the `from-wireguard` path also needs to traverse netfilter, and can't unconditionally redirect straight to the pod.

smagnani96 commented:
/test

smagnani96 commented:
/test

smagnani96 committed:
bpf:wireguard: delivery host packets to bpf_host for ingress policies

This effectively restores HostFw for WireGuard.
Prior to this, we were erroneously always returning to the stack all
packets destined for local host, skipping the HostFw policies if enabled.

With this patch, packets for local host will always be delivered to
cilium_net@egress, similarly to what we do in bpf_overlay after decap.
With HostFw enabled, the to-host program in `cilium_host@ingress` will
then enforce policies.

This patch does not affect packets for local endpoint:
* With BPF Host Routing: will be directly delivered to the pod, tail
  calling into its ep->lxc_id function to enforce ingress policies.
* Without BPF Host Routing: will return to stack, which then goes to its
  to-container installed program to match ingress policies.

Trying to pull-in the whole `host_firewall.h` and `policy.h` would require
to set bpf_wireguard similarly as we do for bpf_host, meaning assigning
an endpoint ID, otherwise host policies would block all host related
packets (ep id == 0). For this reason, we decide here to go through cilium_host.

From a bpf test perspective:
* no changes for packet to/from local endpoint INGRESS/EGRESS
* no changes for packet from local host EGRESS
* packet to local host INGRESS: differently than bpf_host, in WireGuard
                                we always redirect to cilium_host@ingress.

While fixing this bits, let's move the superfluous `revalidate_data`
post NodePort inside the NodePort code, as not needed otherwise.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 commented:
/test
